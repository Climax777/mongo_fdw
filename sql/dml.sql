\set MONGO_HOST			'\'localhost\''
\set MONGO_PORT			'\'27017\''
\set MONGO_USER_NAME	'\'edb\''
\set MONGO_PASS			'\'edb\''

-- Before running this file User must create database mongo_fdw_regress and
-- mongo_fdw_regress1 databases on MongoDB with all permission for 'edb' user
-- with 'edb' password and ran mongodb_init.sh file to load collections.

\c contrib_regression
CREATE EXTENSION IF NOT EXISTS mongo_fdw;
CREATE SERVER mongo_server FOREIGN DATA WRAPPER mongo_fdw
  OPTIONS (address :MONGO_HOST, port :MONGO_PORT);
CREATE USER MAPPING FOR public SERVER mongo_server;

-- Create foreign tables
CREATE FOREIGN TABLE f_mongo_test (_id name, a int, b varchar) SERVER mongo_server
  OPTIONS (database 'mongo_fdw_regress', collection 'mongo_test');
CREATE FOREIGN TABLE f_mongo_test1 (_id name, a int, b varchar) SERVER mongo_server
  OPTIONS (database 'mongo_fdw_regress1', collection 'mongo_test1');
CREATE FOREIGN TABLE f_mongo_test2 (_id name, a int, b varchar) SERVER mongo_server
  OPTIONS (database 'mongo_fdw_regress2', collection 'mongo_test2');
-- Creating foreign table without specifying database.
CREATE FOREIGN TABLE f_mongo_test3 (_id name, a int, b varchar) SERVER mongo_server
  OPTIONS (collection 'mongo_test3');

-- Verify the INSERT/UPDATE/DELETE operations on a collection (mongo_test)
-- exist in a database (mongo_fdw_regress) in mongoDB.
SELECT a,b FROM f_mongo_test ORDER BY 1, 2;
INSERT INTO f_mongo_test VALUES ('0', 10 , 'INSERT');
SELECT a,b FROM f_mongo_test ORDER BY 1, 2;
UPDATE f_mongo_test SET b = 'UPDATE' WHERE a = 10;
SELECT a,b FROM f_mongo_test ORDER BY 1, 2;
DELETE FROM f_mongo_test WHERE a = 10;
SELECT a,b FROM f_mongo_test ORDER BY 1, 2;

-- Verify the INSERT/UPDATE/DELETE operations on a collection (mongo_test1)
-- not exist in a database (mongo_fdw_regress1) in mongoDB.
SELECT a,b FROM f_mongo_test1 ORDER BY 1, 2;
INSERT INTO f_mongo_test1 VALUES ('0', 10 , 'INSERT');
SELECT a,b FROM f_mongo_test1 ORDER BY 1, 2;
UPDATE f_mongo_test1 SET b = 'UPDATE' WHERE a = 10;
SELECT a,b FROM f_mongo_test1 ORDER BY 1, 2;
DELETE FROM f_mongo_test1 WHERE a = 10;
SELECT a,b FROM f_mongo_test1 ORDER BY 1, 2;

-- Verify the INSERT/UPDATE/DELETE operations on a collection (mongo_test2)
-- not exist in a non exist database (mongo_fdw_regress2) in mongoDB.
SELECT a,b FROM f_mongo_test2 ORDER BY 1, 2;
INSERT INTO f_mongo_test2 VALUES ('0', 10 , 'INSERT');
SELECT a,b FROM f_mongo_test2 ORDER BY 1, 2;
UPDATE f_mongo_test2 SET b = 'UPDATE' WHERE a = 10;
SELECT a,b FROM f_mongo_test2 ORDER BY 1, 2;
DELETE FROM f_mongo_test2 WHERE a = 10;
SELECT a,b FROM f_mongo_test2 ORDER BY 1, 2;

-- Verify the INSERT/UPDATE/DELETE operations on a collection (mongo_test)
-- when foreign table created without database option.
SELECT a,b FROM f_mongo_test3 ORDER BY 1, 2;
INSERT INTO f_mongo_test3 VALUES ('0', 10 , 'INSERT');
SELECT a,b FROM f_mongo_test3 ORDER BY 1, 2;
UPDATE f_mongo_test3 SET b = 'UPDATE' WHERE a = 10;
SELECT a,b FROM f_mongo_test3 ORDER BY 1, 2;
DELETE FROM f_mongo_test3 WHERE a = 10;
SELECT a,b FROM f_mongo_test3 ORDER BY 1, 2;

-- FDW-158: Fix server crash when analyzing a foreign table.
SELECT reltuples FROM pg_class WHERE relname = 'f_mongo_test';
ANALYZE f_mongo_test;
-- Should give correct number of rows now.
SELECT reltuples FROM pg_class WHERE relname = 'f_mongo_test';
-- Check count using select query on table.
SELECT count(*) FROM f_mongo_test;

-- Some more variants of vacuum and analyze
VACUUM f_mongo_test;
VACUUM FULL f_mongo_test;
VACUUM FREEZE f_mongo_test;
ANALYZE f_mongo_test;
ANALYZE f_mongo_test(a);
VACUUM ANALYZE f_mongo_test;

-- Cleanup
DROP FOREIGN TABLE f_mongo_test;
DROP FOREIGN TABLE f_mongo_test1;
DROP FOREIGN TABLE f_mongo_test2;
DROP FOREIGN TABLE f_mongo_test3;
DROP USER MAPPING FOR public SERVER mongo_server;
DROP SERVER mongo_server;
DROP EXTENSION mongo_fdw;
