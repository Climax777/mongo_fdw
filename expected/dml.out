\set MONGO_HOST			'\'localhost\''
\set MONGO_PORT			'\'27017\''
\set MONGO_USER_NAME	'\'edb\''
\set MONGO_PASS			'\'edb\''
-- Before running this file User must create database mongo_fdw_regress and
-- mongo_fdw_regress1 databases on MongoDB with all permission for 'edb' user
-- with 'edb' password and ran mongodb_init.sh file to load collections.
\c contrib_regression
CREATE EXTENSION IF NOT EXISTS mongo_fdw;
CREATE SERVER mongo_server FOREIGN DATA WRAPPER mongo_fdw
  OPTIONS (address :MONGO_HOST, port :MONGO_PORT);
CREATE USER MAPPING FOR public SERVER mongo_server;
-- Create foreign tables
CREATE FOREIGN TABLE f_mongo_test (_id name, a int, b varchar) SERVER mongo_server
  OPTIONS (database 'mongo_fdw_regress', collection 'mongo_test');
CREATE FOREIGN TABLE f_mongo_test1 (_id name, a int, b varchar) SERVER mongo_server
  OPTIONS (database 'mongo_fdw_regress1', collection 'mongo_test1');
CREATE FOREIGN TABLE f_mongo_test2 (_id name, a int, b varchar) SERVER mongo_server
  OPTIONS (database 'mongo_fdw_regress2', collection 'mongo_test2');
-- Creating foreign table without specifying database.
CREATE FOREIGN TABLE f_mongo_test3 (_id name, a int, b varchar) SERVER mongo_server
  OPTIONS (collection 'mongo_test3');
-- Verify the INSERT/UPDATE/DELETE operations on a collection (mongo_test)
-- exist in a database (mongo_fdw_regress) in mongoDB.
SELECT a,b FROM f_mongo_test ORDER BY 1, 2;
 a |           b           
---+-----------------------
 0 | mongo_test collection
(1 row)

INSERT INTO f_mongo_test VALUES ('0', 10 , 'INSERT');
SELECT a,b FROM f_mongo_test ORDER BY 1, 2;
 a  |           b           
----+-----------------------
  0 | mongo_test collection
 10 | INSERT
(2 rows)

UPDATE f_mongo_test SET b = 'UPDATE' WHERE a = 10;
SELECT a,b FROM f_mongo_test ORDER BY 1, 2;
 a  |           b           
----+-----------------------
  0 | mongo_test collection
 10 | UPDATE
(2 rows)

DELETE FROM f_mongo_test WHERE a = 10;
SELECT a,b FROM f_mongo_test ORDER BY 1, 2;
 a |           b           
---+-----------------------
 0 | mongo_test collection
(1 row)

-- Verify the INSERT/UPDATE/DELETE operations on a collection (mongo_test1)
-- not exist in a database (mongo_fdw_regress1) in mongoDB.
SELECT a,b FROM f_mongo_test1 ORDER BY 1, 2;
 a | b 
---+---
(0 rows)

INSERT INTO f_mongo_test1 VALUES ('0', 10 , 'INSERT');
SELECT a,b FROM f_mongo_test1 ORDER BY 1, 2;
 a  |   b    
----+--------
 10 | INSERT
(1 row)

UPDATE f_mongo_test1 SET b = 'UPDATE' WHERE a = 10;
SELECT a,b FROM f_mongo_test1 ORDER BY 1, 2;
 a  |   b    
----+--------
 10 | UPDATE
(1 row)

DELETE FROM f_mongo_test1 WHERE a = 10;
SELECT a,b FROM f_mongo_test1 ORDER BY 1, 2;
 a | b 
---+---
(0 rows)

-- Verify the INSERT/UPDATE/DELETE operations on a collection (mongo_test2)
-- not exist in a non exist database (mongo_fdw_regress2) in mongoDB.
SELECT a,b FROM f_mongo_test2 ORDER BY 1, 2;
 a | b 
---+---
(0 rows)

INSERT INTO f_mongo_test2 VALUES ('0', 10 , 'INSERT');
SELECT a,b FROM f_mongo_test2 ORDER BY 1, 2;
 a  |   b    
----+--------
 10 | INSERT
(1 row)

UPDATE f_mongo_test2 SET b = 'UPDATE' WHERE a = 10;
SELECT a,b FROM f_mongo_test2 ORDER BY 1, 2;
 a  |   b    
----+--------
 10 | UPDATE
(1 row)

DELETE FROM f_mongo_test2 WHERE a = 10;
SELECT a,b FROM f_mongo_test2 ORDER BY 1, 2;
 a | b 
---+---
(0 rows)

-- Verify the INSERT/UPDATE/DELETE operations on a collection (mongo_test)
-- when foreign table created without database option.
SELECT a,b FROM f_mongo_test3 ORDER BY 1, 2;
 a | b 
---+---
(0 rows)

INSERT INTO f_mongo_test3 VALUES ('0', 10 , 'INSERT');
SELECT a,b FROM f_mongo_test3 ORDER BY 1, 2;
 a  |   b    
----+--------
 10 | INSERT
(1 row)

UPDATE f_mongo_test3 SET b = 'UPDATE' WHERE a = 10;
SELECT a,b FROM f_mongo_test3 ORDER BY 1, 2;
 a  |   b    
----+--------
 10 | UPDATE
(1 row)

DELETE FROM f_mongo_test3 WHERE a = 10;
SELECT a,b FROM f_mongo_test3 ORDER BY 1, 2;
 a | b 
---+---
(0 rows)

-- FDW-158: Fix server crash when analyzing a foreign table.
SELECT reltuples FROM pg_class WHERE relname = 'f_mongo_test';
 reltuples 
-----------
         0
(1 row)

ANALYZE f_mongo_test;
-- Should give correct number of rows now.
SELECT reltuples FROM pg_class WHERE relname = 'f_mongo_test';
 reltuples 
-----------
         1
(1 row)

-- Check count using select query on table.
SELECT count(*) FROM f_mongo_test;
 count 
-------
     1
(1 row)

-- Some more variants of vacuum and analyze
VACUUM f_mongo_test;
WARNING:  skipping "f_mongo_test" --- cannot vacuum non-tables or special system tables
VACUUM FULL f_mongo_test;
WARNING:  skipping "f_mongo_test" --- cannot vacuum non-tables or special system tables
VACUUM FREEZE f_mongo_test;
WARNING:  skipping "f_mongo_test" --- cannot vacuum non-tables or special system tables
ANALYZE f_mongo_test;
ANALYZE f_mongo_test(a);
VACUUM ANALYZE f_mongo_test;
WARNING:  skipping "f_mongo_test" --- cannot vacuum non-tables or special system tables
-- FDW-226: Fix COPY FROM and foreign partition routing results in a
-- server crash
-- Should fail as foreign table direct copy is not supported
COPY f_mongo_test TO '/tmp/data.txt' delimiter ',';
ERROR:  cannot copy from foreign table "f_mongo_test"
HINT:  Try the COPY (SELECT ...) TO variant.
COPY f_mongo_test (a) TO '/tmp/data.txt' delimiter ',';
ERROR:  cannot copy from foreign table "f_mongo_test"
HINT:  Try the COPY (SELECT ...) TO variant.
COPY f_mongo_test (b) TO '/tmp/data.txt' delimiter ',';
ERROR:  cannot copy from foreign table "f_mongo_test"
HINT:  Try the COPY (SELECT ...) TO variant.
-- Should pass
COPY (SELECT * FROM f_mongo_test) TO '/tmp/data.txt' delimiter ',';
COPY (SELECT a, b FROM f_mongo_test) TO '/tmp/data.txt' delimiter ',';
COPY (SELECT a FROM f_mongo_test) TO '/tmp/data.txt' delimiter ',';
COPY (SELECT b FROM f_mongo_test) TO '/tmp/data.txt' delimiter ',';
-- Should throw an error as copy to foreign table is not supported
DO
$$
BEGIN
  COPY f_mongo_test FROM '/tmp/data.txt' delimiter ',';
EXCEPTION WHEN others THEN
  IF SQLERRM = 'COPY and foreign partition routing not supported in mongo_fdw' OR
     SQLERRM = 'cannot copy to foreign table "f_mongo_test"' THEN
    RAISE NOTICE 'ERROR:  COPY and foreign partition routing not supported in mongo_fdw';
  ELSE
    RAISE NOTICE '%', SQLERRM;
  END IF;
END;
$$
LANGUAGE plpgsql;
NOTICE:  ERROR:  COPY and foreign partition routing not supported in mongo_fdw
DO
$$
BEGIN
  COPY f_mongo_test(a, b) FROM '/tmp/data.txt' delimiter ',';
EXCEPTION WHEN others THEN
  IF SQLERRM = 'COPY and foreign partition routing not supported in mongo_fdw' OR
     SQLERRM = 'cannot copy to foreign table "f_mongo_test"' THEN
    RAISE NOTICE 'ERROR:  COPY and foreign partition routing not supported in mongo_fdw';
  ELSE
    RAISE NOTICE '%', SQLERRM;
  END IF;
END;
$$
LANGUAGE plpgsql;
NOTICE:  ERROR:  COPY and foreign partition routing not supported in mongo_fdw
DO
$$
BEGIN
  COPY f_mongo_test(a) FROM '/tmp/data.txt' delimiter ',';
EXCEPTION WHEN others THEN
  IF SQLERRM = 'COPY and foreign partition routing not supported in mongo_fdw' OR
     SQLERRM = 'cannot copy to foreign table "f_mongo_test"' THEN
    RAISE NOTICE 'ERROR:  COPY and foreign partition routing not supported in mongo_fdw';
  ELSE
    RAISE NOTICE '%', SQLERRM;
  END IF;
END;
$$
LANGUAGE plpgsql;
NOTICE:  ERROR:  COPY and foreign partition routing not supported in mongo_fdw
DO
$$
BEGIN
  COPY f_mongo_test(b) FROM '/tmp/data.txt' delimiter ',';
EXCEPTION WHEN others THEN
  IF SQLERRM = 'COPY and foreign partition routing not supported in mongo_fdw' OR
     SQLERRM = 'cannot copy to foreign table "f_mongo_test"' THEN
    RAISE NOTICE 'ERROR:  COPY and foreign partition routing not supported in mongo_fdw';
  ELSE
    RAISE NOTICE '%', SQLERRM;
  END IF;
END;
$$
LANGUAGE plpgsql;
NOTICE:  ERROR:  COPY and foreign partition routing not supported in mongo_fdw
-- Cleanup
DROP FOREIGN TABLE f_mongo_test;
DROP FOREIGN TABLE f_mongo_test1;
DROP FOREIGN TABLE f_mongo_test2;
DROP FOREIGN TABLE f_mongo_test3;
DROP USER MAPPING FOR public SERVER mongo_server;
DROP SERVER mongo_server;
DROP EXTENSION mongo_fdw;
